prov/gni/src/gnix_mr.c

```diff
@@ -195,23 +364,342 @@ static int fi_gnix_mr_close(fid_t fid)
+	}
+
+	return FI_SUCCESS;
 }
 
+/**
+ * Function to register memory with the cache
+ *
+ * @param[in] cache        gnix memory registration cache pointer
+ * @param[in] mr           gnix memory region descriptor pointer
+ * @param[in] domain       gnix domain pointer
+ * @param[in] address      base address of the memory region to be registered
+ * @param[in] length       length of the memory region to be registered
+ * @param[in] dst_cq_hndl  destination gni cq handle for cq event delivery
+ * @param[in] flags        gni memory registration flags
+ * @param[in] vmdh_index   desired index for the new vmdh
+ * @param[in,out] mem_hndl gni memory handle pointer to written to and returned
+ */
+static int __mr_cache_register(
+		gnix_mr_cache_t          *cache,
+		struct gnix_fid_mem_desc *mr,
+		struct gnix_fid_domain   *domain,
+		uint64_t                 address,
+		uint64_t                 length,
+		gni_cq_handle_t          dst_cq_hndl,
+		uint32_t                 flags,
+		uint32_t                 vmdh_index,
+		gni_mem_handle_t         *mem_hndl)
+{
+	RbtStatus rc;
+	RbtIterator iter;
+	gnix_mr_cache_key_t key, *e_key;
+	gnix_mr_cache_entry_t *entry;
+	struct gnix_nic *nic;
+	gni_return_t grc;
+
+	/* if we shouldn't introduce any new elements, return -FI_ENOSPC */
+	if (atomic_get(&cache->inuse_elements) >= cache->attr.hard_reg_limit &&
+			cache->attr.hard_reg_limit > 0)
+		return FI_ENOSPC;
+
+	/* build key for searching */
+	key.address = address;
+	key.length = length;
+
+	/* Is the key in the inuse tree? */
+	iter = rbtFind(cache->inuse, &key);
+	if (iter) {
+		/* Let's increment the ref count of the entry */
+		rbtKeyValue(cache->inuse, iter, (void **) &e_key,
+				(void **) &entry);
+
+		__mr_cache_entry_get(cache, entry);
+		nic = entry->nic;
+
+		goto success;
+	} else if (cache->attr.lazy_deregistration) {
+		/* if lazy deregistration is in use, we can check the
+		 *   stale tree
+		 */
+		iter = rbtFind(cache->stale, &key);
+		if (iter) {
+			rbtKeyValue(cache->stale, iter, (void **) &e_key,
+					(void **) &entry);
+
+			/* reset the reference count as it should be zero from
+			 *   being in the stale tree anyway
+			 */
+			atomic_set(&entry->ref_cnt, 1);
+
+			/* clear the element from the stale cache */
+			rbtErase(cache->stale, iter);
+			atomic_dec(&cache->stale_elements);
+
+			rc = rbtInsert(cache->inuse, (void *) &e_key,
+					(void *) entry);
+			if (rc == RBT_STATUS_MEM_EXHAUSTED) {
+				__mr_cache_entry_destroy(entry);
+				return -FI_ENOMEM;
+			}
+
+
+			goto success;
+		}
+	}
+
+	/* if we made it here, we didn't find the entry at all */
+	entry = calloc(1, sizeof(*entry));
+	if (!entry)
+		return -FI_ENOMEM;
+
+	/* TODO: should we just try the first nic we find? */
+	list_for_each(&domain->nic_list, nic, list)
+	{
+		grc = GNI_MemRegister(nic->gni_nic_hndl, address, length,
+					dst_cq_hndl, flags,
+					vmdh_index, &entry->mem_hndl);
+		if (grc == GNI_RC_SUCCESS)
+			break;
+	}
```

Comment:
The code doesn't actually register on every nic, just the first one it can register on successfully. The disadvantage to the loop, is that if it can't register the memory at all, it will try on every nic.